### DISCUSSION TITLE
320x200 aspect ratio

### DISCUSSION THREAD

**Discussion by geist22** [swfans/syndwarsfx#182] [General]
Not sure if this qualifies as an issue, so I wanted to open a discussion first.

Currently, content with a 320x200 resolution displays in a 16:10 aspect ratio, which assumes square pixels. As far as I understand it, 320x200 was generally (with some exceptions) used as a 4:3 resolution with non-quare pixels. This also seems to be the general consensus that retro game communities such as DOSBox Staging or DOOM have arrived at.

The following things are affected in swars port right now:
- intro FMV
- ending FMV (agent on the moon)
- ending screen with text (after ending FMV)
- low-res gameplay

The FMV for the orbital lift is unaffected.

I think at least the low-res gameplay and ending screen should be in 4:3. The FMVs probably too, though I'm not so sure. The FMVs look like they already have black bars, not to fill the screen but integrated in the videos themselves. I've made screenshots of the Bullfrog logo in the intro in both 4:3 and 16:10 and compared them to the logo found in printed materials such as the manual, and the one in 16:10 matches while the 4:3 is too narrow. This might be a hint, or just an oversight like the Imp's fireball in DOOM, which is a circle in 16:10 but oval in the intended resolution of 4:3. I've tried to find material such as old game magazines with screenshots of the intro to see what they use. Haven't found much, though 4:3 is used in some trailers that show the intro.

What do you think? Personally, I believe everything 320x200 should just be in 4:3, just like 640x480.

If you're interested, here's a video concerning aspect ratios of DOS games with some examples:
https://www.youtube.com/watch?v=YvckyWxHAIw

**Comment by mefistotelis** [comment]
For the in-game 320x200 the game just uses the same angle computation for 3D scene, and at least some of the same sprites (panel sprites are different, but they don't seem to be thinner and longer, just half the size). So the fact that pixels are non-square is just ignored, on non-digital displays everything is stretched vertically compared to 640x480.

I'm not exactly sure if I should care about this aspect more than Bullfrog did.

For 320x200 movies - this is not going to be fixed in software, but I suspect we will at some point fix it by providing better quality video file. The file could be cropped to have black frames removed, and adjusted for square pixels.

I'm not sure if the code would then scale the video to fill one of the dimensions, or only multiply the size (2x/3x/4x...) and center on screen. We'll decide when we're there.

**Reply by geist22** [reply]
> I'm not exactly sure if I should care about this aspect more than Bullfrog did.

I don't think Bullfrog had to care about this at all - monitors back then were usually 4:3, so everything was just displayed in 4:3. This problem only started to appear with newer monitors and square pixels. Most ports and emulators use aspect ratio correction to stretch 320x200 and display it in 4:3.

Though it's most likely beyond my abilities, I might take a look and see if I could implement some form of aspect ratio correction here.

**Reply by mefistotelis** [reply]
> I don't think Bullfrog had to care about this at all - monitors back then were usually 4:3, so everything was just displayed in 4:3.

Right, that's true.
What I meant is that Bullfrog used the same drawing perspective for 3D world in 320x200 and for 640x480. Regardless of display size, these two resolutions have different W/H ratio, but still the game does not compensate for that, and even panel sprites do not seem to be altered to take this ratio difference into account.

> I might take a look and see if I could implement some form of aspect ratio correction

Ok, we could then judge whether it helps.

As I mentioned before, I currently imagine a fix as just updated version of low-res videos - maybe rescaled to 640x480 with some AI filter, and the black borders removed so it can easier fit a modern screen.

For me the target in regard to display is that the port will auto-configure to use native display resolution (the borderless mode).

**Reply by geist22** [reply]
> What I meant is that Bullfrog used the same drawing perspective for 3D world in 320x200 and for 640x480. Regardless of display size, these two resolutions have different W/H ratio, but still the game does not compensate for that, and even panel sprites do not seem to be altered to take this ratio difference into account.

You're right, it just did some comparison:

| DOSBox            |  Port |
:-------------------------:|:-------------------------:
![dosbox640](https://github.com/user-attachments/assets/abec9004-60bc-4527-8227-bcd1b17d76a3) | ![port640](https://github.com/user-attachments/assets/0aec8ed6-0c07-433f-b9c3-c6a30ff98634)
![dosbox320](https://github.com/user-attachments/assets/8f8c70ca-0ef1-4c13-ac91-48b861930086) | ![port320](https://github.com/user-attachments/assets/0e999cee-9a3e-4677-a821-76ba637f02d9)

With that we can conclude that the aspect ratio of the game world in 320x200 was wrong in 4:3 and works correctly in the port. That would leave the intro, outro and ending screen.

| DOSBox            |  Port |
:-------------------------:|:-------------------------:
![endingdosbox](https://github.com/user-attachments/assets/ca432aa0-d900-4e6b-be87-4a2105d63a94) | ![endingport](https://github.com/user-attachments/assets/4403376b-3280-444e-a0f9-66700d535b30)

Stretched version looked a bit wide to me, but when looking at the agent's face, it looks a bit narrow in 4:3, does it not? The glow around the star you can see during the ending video also resembles more of a round circle in the 16:9 version compared to 4:3.

Though I don't have a comparison now, I always thought the people and agents in the intro movie looked a bit fat in 16:9, but the comparison of the Bullfrog logo I did previously could speak otherwise.

Maybe SWARS really is one of the rare exceptions where 320x200 was "wrong" in 4:3?

> For me the target in regard to display is that the port will auto-configure to use native display resolution (the borderless mode).

Borderless fullscreen would be great, is it implemented yet? I've been using a 1280x960 window as it fits a 1920x1080 display and is also exactly double of 640x480 so it is easy to scale. Some programs causes a major annoyance when using lower resolutions in fullscreen (such as moving and resizing all open windows and moving around items on the desktop), and the SWARS port used to be one of them at least in the vexillium days.
